airbyte-integrations/connectors/source-asana/source_asana/streams.py

```diff
@@ -2,23 +2,37 @@
 # Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 #
 
-from __future__ import annotations
 
 from abc import ABC
-from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Type
+from typing import Any, Iterable, Mapping, MutableMapping, Optional
 
 import requests
 from airbyte_cdk.models import SyncMode
 from airbyte_cdk.sources.streams.http import HttpStream
 
+ASANA_ERRORS_MAPPING = {
+    402: "Stream is available to premium organizations and workspaces",
+    403: "Not enough permissions",
+    404: "Object specified by the request does not exist",
+    451: "This request was blocked for legal reasons",
+}
+
 
 class AsanaStream(HttpStream, ABC):
     url_base = "https://app.asana.com/api/1.0/"
-
     primary_key = "gid"
-
     # Asana pagination could be from 1 to 100.
     page_size = 100
+    raise_on_http_errors = True
+
+    def should_retry(self, response: requests.Response) -> bool:
+        if response.status_code in ASANA_ERRORS_MAPPING.keys():
+            self.logger.error(
+                f"Skipping stream {self.name}. {ASANA_ERRORS_MAPPING.get(response.status_code)}. Errors: {response.json().get('errors')}"
```

Comment:
response.json() is not safe, please replace it with `response.text`

Reply:
Why?

Reply:
because it raises an error if a response body is not a json which is very common for 4xx statuses

Reply:
So you suggest converting `string` to JSON instead? or parsing the string to find the errors there? 
Because Asana API always returns the JSON body along with the error messages.

Reply:
If you're sure the response will always contain a valid json within this `if` block, feel free to leave it as is. Otherwise U suggest logging the raw response.text.

Regarding all the other connectors - ideally yes, this should be done for all of them because it's a common antipattern